src/arch/southern-islands/timing/ComputeUnit.cc

```diff
@@ -554,13 +559,29 @@ void ComputeUnit::RemoveWorkGroup(WorkGroup *work_group)
 			index,
 			work_group->id_in_compute_unit);
 
-	// Erase work group                                                      
+	// Unmap work group from the compute unit
 	assert(work_group->compute_unit_work_groups_iterator != 
 			work_groups.end());
 	work_groups[work_group->id_in_compute_unit] = nullptr;
 }
 
 
+void ComputeUnit::Reset()
+{
+	// Save timing simulator
+	timing = Timing::getInstance();
+
+	// Debug info
+	Emulator::scheduler_debug << misc::fmt("@%lld compute unit %d "
+			"reset\n",
+			timing->getCycle(),
+			index);
+
+	// Reset the workgroups size to 0
+	work_groups.resize(0);
```

Comment:
- Notice the difference between `std::vector::clear()` and `std::vector::resize(0)`. Even though in the former has a subset of the functionality of the latter, I think that `clear()` is a more appropriate function to invoke. In general, we don't want to mess with the internal size that the C++ STL devotes to the vector container.

- But... is this really what we want to do? If the `work_groups` vector has something in it, it is probably because something is still going on in the compute unit. Is it safe to just reset this data structure? Doesn't it conflict with the state of other data structures? Isn't the vector of work-groups progressively emptied as work-groups are drained? You need to be very careful emptying structures that contain unique pointers, since this involve freeing the objects they point to. No other container should have regular pointers to them at this point.